frontend/apps/reader/modules/readerbookmark.lua

```diff
@@ -215,7 +215,38 @@ function ReaderBookmark:gotoBookmark(pn_or_xp)
     end
 end
 
+-- This function adds "chapter" property to highlights already saved in the document
+function ReaderBookmark:updateHighlightsIfNeeded()
+    local version = self.ui.doc_settings:readSetting("bookmarks_version") or 0
+    if version >= 20200511 then
+        return
+    end
+
+    for _, highlights in pairs(self.view.highlight.saved) do
+        for _, highlight in pairs(highlights) do
+            local pg_or_xp = self.ui.document.info.has_pages and
+                    highlight.pos0.page or highlight.pos0
+            local chapter_name = self.ui.toc:getTocTitleByPage(pg_or_xp)
+            highlight.chapter = chapter_name
+        end
+    end
+
+    for _, bookmark in ipairs(self.bookmarks) do
+        if bookmark.pos0 then
+            local pg_or_xp = self.ui.document.info.has_pages and
+                    bookmark.pos0.page or bookmark.pos0
+                local chapter_name = self.ui.toc:getTocTitleByPage(pg_or_xp)
+            bookmark.chapter = chapter_name
+        elseif bookmark.page then -- dogear bookmark
```

Comment:
Also `bookmark.pos0.page` with PDF:
```lua
    ["bookmarks"] = {
        [4] = {
            ["page"] = 5,
            ["highlighted"] = true,
            ["notes"] = "T steps. We use the T simulation",
            ["datetime"] = "2020-02-27 12:19:58",
            ["pos1"] = {
                ["page"] = 5,
                ["x"] = 354.14966868083,
                ["y"] = 278.53742190043,
                ["rotation"] = 0,
                ["zoom"] = 1.5473683825295
            },
            ["pos0"] = {
                ["page"] = 5,
                ["x"] = 222.31293070475,
                ["y"] = 281.12245597839,
                ["rotation"] = 0,
                ["zoom"] = 1.5473683825295
            }
        },
        [1] = {
            ["page"] = 8,
            ["highlighted"] = true,
            ["notes"] = "converges uniformly on compact sets forthe case of random walk Metropolis algorithm with proposal",
            ["datetime"] = "2020-05-24 17:06:49",
            ["pos1"] = {
                ["y"] = 529.75777777778,
                ["x"] = 349.50444444444
            },
            ["pos0"] = {
                ["y"] = 520.07150326797,
                ["x"] = 296.26222222222
            }
        },
```
The highlight I just made today has no `pos0.page` - while one I made some months ago has one... (no idea why, don't wanna investigate :) Our highlights are really a mess...)
But they both have a `page`

Reply:
Ok, that doesn't make sense :)
I checked my highlights in pdfs and all had `pos0.page` property, that's why I did it like that. I'll fix them soon. Actually, I'm not sure if my highlights had `page` property, I'll need to investigate some more.

Reply:
Hopefully I fixed it, it's probably some old version, that didn't add `page` to `pos0`. Could you confirm, that doesn't happen with highlights? I tried to replicate, but couldn't. I took a look at v2019.01 `readerhighlight.lua`/`readerbookmark.lua` and it seems it was already adding `pos0.page`. This is probably a result of some ancient version.

Reply:
> it's probably some old version, that didn't add page to pos0

Well, the bookmark that didn't have pos0.page was made the day I posted that comment :)
`["datetime"] = "2020-05-24 17:06:49",`
(don't have time to look at the code what may make it add it and what not)

> Could you confirm, that doesn't happen with highlights?

In my other comment above, https://github.com/koreader/koreader/pull/6146#discussion_r429646441 , it looks like I saw it happening with highlights too.